src/function/scalar/list/list_concat.cpp

```diff
@@ -0,0 +1,132 @@
+		                   rhs_entry.offset);
+	}
+	D_ASSERT(ListVector::GetListSize(result) == offset);
+
+	if (lhs.GetVectorType() == VectorType::CONSTANT_VECTOR && rhs.GetVectorType() == VectorType::CONSTANT_VECTOR) {
+		result.SetVectorType(VectorType::CONSTANT_VECTOR);
+	}
+}
+
+static unique_ptr<FunctionData> ListConcatBind(ClientContext &context, ScalarFunction &bound_function,
+                                               vector<unique_ptr<Expression>> &arguments) {
+	D_ASSERT(bound_function.arguments.size() == 2);
+
+	auto &lhs = arguments[0]->return_type;
+	auto &rhs = arguments[1]->return_type;
+	if (lhs.id() == LogicalTypeId::SQLNULL) {
+		bound_function.arguments[0] = LogicalType::SQLNULL;
+	}
+	if (rhs.id() == LogicalTypeId::SQLNULL) {
+		bound_function.arguments[1] = LogicalType::SQLNULL;
+	}
+	if (lhs.id() == LogicalTypeId::SQLNULL || rhs.id() == LogicalTypeId::SQLNULL) {
+		bound_function.return_type = LogicalType::SQLNULL;
+	} else {
+		D_ASSERT(lhs.id() == LogicalTypeId::LIST);
+		D_ASSERT(rhs.id() == LogicalTypeId::LIST);
+
+		auto &lhs_child_type = ListType::GetChildType(lhs);
+		auto &rhs_child_type = ListType::GetChildType(rhs);
+		if (lhs_child_type.id() == LogicalTypeId::SQLNULL) {
+			bound_function.arguments[0] = rhs;
+			bound_function.arguments[1] = rhs;
+			bound_function.return_type = rhs;
+		} else if (rhs_child_type.id() == LogicalTypeId::SQLNULL) {
+			bound_function.arguments[0] = lhs;
+			bound_function.arguments[1] = lhs;
+			bound_function.return_type = lhs;
+		} else if (lhs_child_type.id() != rhs_child_type.id()) {
+			throw TypeMismatchException(lhs, rhs, "Cannot concatenate lists with different child types");
+		} else {
+			bound_function.arguments[0] = lhs;
+			bound_function.arguments[1] = lhs;
+			bound_function.return_type = lhs;
+		}
+	}
+	return make_unique<VariableReturnBindData>(bound_function.return_type);
+}
+
+static unique_ptr<BaseStatistics> ListConcatStats(ClientContext &context, BoundFunctionExpression &expr,
+                                                  FunctionData *bind_data,
+                                                  vector<unique_ptr<BaseStatistics>> &child_stats) {
+	D_ASSERT(child_stats.size() == 2);
+	if (!child_stats[0] || !child_stats[1]) {
+		return nullptr;
+	}
+
+	auto &left_stats = (ListStatistics &)*child_stats[0];
+	auto &right_stats = (ListStatistics &)*child_stats[1];
+
+	auto stats = left_stats.Copy();
+	stats->Merge(right_stats);
+
+	return stats;
+}
+
+void ListConcatFun::RegisterFunction(BuiltinFunctions &set) {
+	// the arguments and return types are actually set in the binder function
+	ScalarFunction lfun({LogicalType::LIST(LogicalType::ANY), LogicalType::LIST(LogicalType::ANY)},
+	                    LogicalType::LIST(LogicalType::ANY), ListConcatFunction, false, ListConcatBind, nullptr,
+	                    ListConcatStats);
+	set.AddFunction({"list_concat", "list_cat", "array_concat", "array_cat"}, lfun);
+}
```

Comment:
Postgres also overloads [`||` for lists ](https://www.postgresql.org/docs/9.1/functions-array.html), perhaps we should allow the same.